src/python/zquantum/core/wip/circuit/_circuit.py

```diff
@@ -0,0 +1,148 @@
+            return False
+
+        for i in range(len(self.gates)):
+            if self.gates[i] != another_circuit.gates[i]:
+                return False
+
+        # TODO: figure out if this check is redundant when reworking Circuit
+        if len(self.symbolic_params) != len(another_circuit.symbolic_params):
+            return False
+
+        return True
+
+    def __add__(self, other_circuit):
+        """Add two circuits.
+        """
+        new_circuit = type(self)()
+        new_circuit.gates = self.gates + other_circuit.gates
+        return new_circuit
+
+    def evaluate(self, symbols_map: Dict):
+        """ Create a copy of the current Circuit with the parameters of each gate evaluated to the values 
+        provided in the input symbols map
+
+        Args:
+            symbols_map (Dict): A map of the symbols/gate parameters to new values
+        """
+        circuit_class = type(self)
+        evaluated_gate_list = [gate.evaluate(symbols_map) for gate in self.gates]
+        evaluated_circuit = circuit_class(gates=evaluated_gate_list)
+        return evaluated_circuit
+
+    def to_dict(self, serializable: bool = True):
+        """ Creates a dictionary representing a circuit.
+
+        Args:
+            serializable (bool): If true, the returned dictionary is serializable so that it can be stored
+                in JSON format
+        Returns:
+            Dict: keys are schema, qubits, gates, and symbolic_params
+        """
+        circuit_dict = {"schema": SCHEMA_VERSION + "-circuit"}
+        if serializable:
+            circuit_dict["qubits"] = list(self.qubits)
+            circuit_dict["gates"] = [
+                gate.to_dict(serializable=True) for gate in self.gates
+            ]
+            circuit_dict["symbolic_params"] = [
+                str(param) for param in self.symbolic_params
+            ]
+        else:
+            circuit_dict["qubits"] = self.qubits
+            circuit_dict["gates"] = [
+                gate.to_dict(serializable=False) for gate in self.gates
+            ]
+            circuit_dict["symbolic_params"] = self.symbolic_params
+        return circuit_dict
+
+    def save(self, filename: str):
+        """ Save the Circuit object to file in JSON format
+
+        Args:
+            filename (str): The path to the file to store the Circuit
+        """
+        with open(filename, "w") as f:
+            f.write(json.dumps(self.to_dict(serializable=True), indent=2))
+
+    @classmethod
+    def load(cls, data: Union[Dict, TextIO]):
+        """ Load a Circuit object from either a file/file-like object or a dictionary
+
+        Args:
+            data (Union[Dict, TextIO]): The data to load into the Circuit object
+
+        Returns:
+            Circuit
+        """
+        if isinstance(data, str):
+            with open(data, "r") as f:
+                data = json.load(f)
+        elif not isinstance(data, dict):
+            data = json.load(data)
```

Comment:
What kind of types would this be applicable for?

Reply:
I plan to reimplement this as well, especially that the discussion about data serialization in Orquestra is still ongoing. I'll make sure the type dispatch is more explicit there.

Sneak peek: IMHO the serialization/deserialization should be decoupled from the data structure, with optional convenience API on top of it if users really like `Circuit.load` and `Circuit.save` helpers.